frontend/src/metabase/lib/utils.js

```diff
@@ -135,36 +135,63 @@ const MetabaseUtils = {
     return JSON.parse(JSON.stringify(a));
   },
 
-  // this should correctly compare all version formats Metabase uses, e.x.
-  // 0.0.9, 0.0.10-snapshot, 0.0.10-alpha1, 0.0.10-rc1, 0.0.10-rc2, 0.0.10-rc10
-  // 0.0.10, 0.1.0, 0.2.0, 0.10.0, 1.1.0
-  compareVersions(aVersion, bVersion) {
-    if (!aVersion || !bVersion) {
-      return null;
-    }
-
+  /**
+   * Converts a metabase version to a list of numeric components, it converts pre-release
+   * components to numbers and pads the numeric part to 4 numbers to make comparison easier
+   * @param {string} version
+   * @returns {number[]}
+   */
+  versionToNumericComponents(version) {
     const SPECIAL_COMPONENTS = {
       snapshot: -4,
       alpha: -3,
       beta: -2,
       rc: -1,
     };
 
-    const getComponents = x =>
+    const padRightWithZeros = (array, minLength = 4) => {
+      return Array.from({ length: minLength }).map((_, i) => array[i] || 0);
+    };
+
+    const [versionNumbers, labels = []] =
       // v1.2.3-BETA1
-      x
+      version
         .toLowerCase()
         // v1.2.3-beta1
         .replace(/^v/, "")
         // 1.2.3-beta1
-        .split(/[.-]*([0-9]+)[.-]*/)
-        .filter(c => c)
-        // ["1", "2", "3", "beta", "1"]
-        .map(c => SPECIAL_COMPONENTS[c] || parseInt(c, 10));
-    // [1, 2, 3, -2, 1]
-
-    const aComponents = getComponents(aVersion);
-    const bComponents = getComponents(bVersion);
+        .split("-")
+        // ['1.2.3', 'beta1']
+        .map(
+          substring =>
+            substring
+              .split(/[.-]*([0-9]+)[.-]*/)
+              .filter(c => c)
+              // [["1", "2", "3",] ["beta", "1"]]
+              .map(c => SPECIAL_COMPONENTS[c] || parseInt(c, 10)),
+          // [[1, 2, 3,] [-2, 1]]
+        );
+
+    return [...padRightWithZeros(versionNumbers), ...labels];
```

Comment:
Not very clear to me. What if we rewrite this function to this:
```
const regex = /v?(?<major>\d+)\.?(?<minor>\d+)?\.?(?<patch>\d+)?-?(?<label>\D+)?(?<build>\d+)?/;
const { major = 0, minor = 0, patch = 0, label, build = 0 } =  regex.exec(version);
return [major, minor, patch, SPECIAL_COMPONENTS[label] ?? 0, build];
```

Reply:
done: https://github.com/metabase/metabase/pull/33671/commits/f54789f460cf4b00be76e0ffc545ad005cea6a50

Had to add the ee/oss field and parseInt logic but still seems to be readable